driver/modern_bpf/add_new_syscalls.md

```diff
@@ -0,0 +1,530 @@
+  * As the doc says https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/helpers/store/auxmap_store_params.h#L327-L340
+  * to store a `charbuf` param, so one between `PT_CHARBUF`, `PT_FSPATH`, `PT_FSRELPATH`, we need to use
+  * the `auxmap__store_charbuf_param` helper.
+  * - `MAX_PATH` is the maximum length we want to read from the pathname pointer
+  * - `USER` means that the provided pointer points to userspace memory and not kernel memory.
+  */
+
+ /* Parameter 2: pathname (type: PT_FSPATH) */
+ unsigned long path_pointer = extract__syscall_argument(regs, 0);
+ auxmap__store_charbuf_param(auxmap, path_pointer, MAX_PATH, USER);
+
+ /*=============================== COLLECT PARAMETERS  ===========================*/
+
+ auxmap__finalize_event_header(auxmap);
+
+ auxmap__submit_event(auxmap);
+
+ return 0;
+}
+```
+
+That's all for the exit event implementation!
+
+#### 2. How to call our BPF program
+
+Like before we have to add our entry to the [event_prog_names](https://github.com/falcosecurity/libs/blob/475333b5bfa828209023f0c5bf93330e48f6a46c/userspace/libpman/src/events_prog_names.h#L24) table :point_down:
+
+```c
+[PPME_SYSCALL_ACCESS_X] = "access_x",
+```
+
+#### 4. Write a test to assert the BPF program
+
+We can use the same template seen before for the enter event. Note that the event test template doesn't change according to the approach we use kernel-side, the userspace doesn't know if we are using the `ringbuf` approach or the `auxmap` one.
+
+```cpp
+TEST(SyscallExit, accessX)
+{
+ auto evt_test = get_syscall_event_test(__NR_access, EXIT_EVENT);
+
+ evt_test->enable_capture();
+
+ /*=============================== TRIGGER SYSCALL  ===========================*/
+
+ int32_t mode = W_OK;
+ char pathname[] = "//**null-file-path**//";
+ assert_syscall_state(SYSCALL_FAILURE, "access", syscall(__NR_access, pathname, mode));
+ /* here the syscall fails so the first parameter we assert should be equal
+  * to `-errno`.
+  */
+ int64_t errno_value = -errno;
+
+ /*=============================== TRIGGER SYSCALL ===========================*/
+
+ evt_test->disable_capture();
+
+ evt_test->assert_event_presence();
+
+ if(HasFatalFailure())
+ {
+  return;
+ }
+
+ evt_test->parse_event();
+
+ evt_test->assert_header();
+
+ /*=============================== ASSERT PARAMETERS  ===========================*/
+
+ /* Parameter 1: res (type: PT_ERRNO)*/
+ /* For all the numeric parameters we have the same helper `assert_numeric_param` */
+ evt_test->assert_numeric_param(1, (int64_t)errno_value);
+
+ /* For a `PT_FSPATH` param we have the related helper
+  * https://github.com/falcosecurity/libs/blob/master/test/modern_bpf/event_class/event_class.h#L364-L376
+  */
+ /* Parameter 2: path (type: PT_FSPATH)*/
+ evt_test->assert_charbuf_param(2, pathname);
+
+ /*=============================== ASSERT PARAMETERS  ===========================*/
+
+ /* the exit event has 2 parameters */
+ evt_test->assert_num_params_pushed(2);
+}
+```
+
```

Comment:
```suggestion
## Handling syscalls with complex parameters

Some syscalls' parameters behave differently in the error case. For instance, consider `read(2)`:

#include <unistd.h>

ssize_t read(int fd, void *buf, size_t count);


`read` behaves differently when an error occurs:


RETURN VALUE
       On  success,  the number of bytes read is returned (zero indicates end of file), and the file position is advanced by this number.  It is not
       an error if this number is smaller than the number of bytes requested; this may happen for example because fewer bytes are actually available
       right  now (maybe because we were close to end-of-file, or because we are reading from a pipe, or from a terminal), or because read() was in‐
       terrupted by a signal.  See also NOTES.

       On error, -1 is returned, and errno is set appropriately.  In this case, it is left unspecified whether the file position (if any) changes.


Writing a driver for this syscall is a bit trickier than usual. In the success case, the
exit probe should return the number of bytes read as well as the bytes themselves through the `*buf` parameter. In the error case, the exit probe should return `errno` and nothing/empty for `*buf`.

To verify proper driver handling for both error and success cases, two tests should be created:

### Complex parameter probe

SEC("tp_btf/sys_exit")
int BPF_PROG(read_x,
             struct pt_regs *regs,
             long ret)
{
// ...

/*=============================== COLLECT PARAMETERS  ===========================*/

/* Parameter 1: res (type: PT_ERRNO) */
auxmap__store_s64_param(auxmap, ret);

if(ret > 0)
{
	/* SUCCESS CASE - push a param containing the read buffer */

	/* Parameter 2: data (type: PT_BYTEBUF) */
        unsigned long data_pointer = extract__syscall_argument(regs, 1);
        auxmap__store_bytebuf_param(auxmap, data_pointer, bytes_to_read, USER);
}
else
{
	/* ERROR CASE - push an empty param */

        /* Parameter 2: data (type: PT_BYTEBUF) */
        auxmap__store_empty_param(auxmap);
}

/*=============================== COLLECT PARAMETERS  ===========================*/

// ...
}


### Complex parameter test (Success case)

TEST(SyscallExit, readX)
{
// ...

/*=============================== TRIGGER SYSCALL  ===========================*/

/* Open /dev/urandom for reading */
int fd = syscall(__NR_open, "/dev/urandom", O_RDONLY);
assert_syscall_state(SYSCALL_SUCCESS, "open", fd, NOT_EQUAL, -1);

/* Read data from /dev/urandom */
const unsigned data_len = 64;
char buf[data_len];
ssize_t read_bytes = syscall(__NR_read, fd, (void *)buf, data_len);
assert_syscall_state(SYSCALL_SUCCESS, "read", read_bytes, NOT_EQUAL, -1);

/* Close /dev/urandom fd */
assert_syscall_state(SYSCALL_SUCCESS, "close", syscall(__NR_close, fd), NOT_EQUAL, -1);

/*=============================== TRIGGER SYSCALL ===========================*/

// ...

/*=============================== ASSERT PARAMETERS  ===========================*/

/* Parameter 1: res (type: PT_ERRNO) */
evt_test->assert_numeric_param(1, (int64_t)read_bytes);

/* Parameter 2: data (type: PT_BYTEBUF) */
evt_test->assert_only_param_len(2, (uint32_t)read_bytes);

/*=============================== ASSERT PARAMETERS  ===========================*/

// ...
}


### Complex parameter test (Failure case)

TEST(SyscallExit, readXfailure)
{
// ...

/*=============================== TRIGGER SYSCALL  ===========================*/

/* Read data from an invalid fd */
const unsigned data_len = 64;
char buf[data_len];
ssize_t read_bytes = syscall(__NR_read, -1, (void *)buf, data_len);
int errno_value = -errno;
assert_syscall_state(SYSCALL_FAILURE, "read", read_bytes);

/*=============================== TRIGGER SYSCALL ===========================*/

// ...

/*=============================== ASSERT PARAMETERS  ===========================*/

/* Parameter 1: res (type: PT_ERRNO) */
evt_test->assert_numeric_param(1, (int64_t)errno_value);

/* Parameter 2: data (type: PT_BYTEBUF) */
evt_test->assert_empty_param(2);

/*=============================== ASSERT PARAMETERS  ===========================*/

// ...
}

```

Reply:
thank you for the suggestion ;) I will try to create version 2 of this document ASAP :)